### PR TITLE
Reconnect on error

### PR DESCRIPTION
**Pull request by quatrix**
currently when a tcp socket closes, the exception is unhandled and statsd crashed, to work around i'm catching all errors and also trying to reconnect endlessly in case reimann goes down and then up again. please review

### CODE REVIEW

File: lib/riemann.js [hunk 1]
```diff
@@ -9,8 +9,36 @@ var riemann = require("riemann");
...
  * transport, flush, packet
  */
+
+
 function Riemann(startupTime, config, emitter) {
   this.config = config.riemann;
+  this.emitter = emitter;
+
+  if (this.config.transport && this.config.transport === "tcp") {
+    setInterval(this.reconnectIfNeeded.bind(this), this.config.reconnectInterval);
```
Comment (joshgummersall) [inline lib/riemann.js:19]: It would be a good idea to ensure that `reconnectInterval` is defined as well. That would allow the end user to only enable this functionality if desired.

File: lib/riemann.js [hunk 2]
```diff
@@ -9,8 +9,36 @@ var riemann = require("riemann");
...
+  } else {
+    this.setupRiemannClient();
+  }
+}
+
+Riemann.prototype.reconnectIfNeeded = function() {
+    if (!this.client) {
```
Comment (joshgummersall) [inline lib/riemann.js]: 2 spaces for tabs in this function.

File: lib/riemann.js [hunk 3]
```diff
@@ -9,8 +9,36 @@ var riemann = require("riemann");
...
+
+
 function Riemann(startupTime, config, emitter) {
   this.config = config.riemann;
+  this.emitter = emitter;
+
+  if (this.config.transport && this.config.transport === "tcp") {
+    setInterval(this.reconnectIfNeeded.bind(this), this.config.reconnectInterval);
+  } else {
+    this.setupRiemannClient();
```
Comment (joshgummersall) [inline lib/riemann.js]: I would pull this outside an else block so it is fired immediately on startup, i.e.

``` javascript
if (this.config.transport && this.config.transport === "tcp") {
  setTimeout(...);
}
this.setupRiemannClient();
```

File: lib/riemann.js [hunk 4]
```diff
@@ -9,8 +9,36 @@ var riemann = require("riemann");
...
+}
+
+Riemann.prototype.reconnectIfNeeded = function() {
+    if (!this.client) {
+        if (this.config.debug) {
+            console.log('Connecting to Reimann');
+        }
+
+        this.setupRiemannClient();
+    } else if (this.client.tcp.socket.readyState == 'closed') {
```
Comment (joshgummersall) [inline lib/riemann.js]: It would probably be a good idea to protect this lookup incase any of the components are undefined.

File: lib/riemann.js [hunk 5]
```diff
@@ -32,11 +59,7 @@ Riemann.prototype.onConnect = function () {
 };
 
 Riemann.prototype.onError = function (err) {
-  if (this.config.debug) {
-    console.error(err);
-  } else {
-    throw err;
-  }
+    console.error("Socket error:", err);
```
Comment (joshgummersall) [inline lib/riemann.js]: two spaces for tabs.

File: lib/riemann.js [hunk 6]
```diff
@@ -113,11 +136,17 @@ Riemann.prototype.handlePacketEvent = function (eventString) {
 Riemann.prototype.send = function (eventData) {
   var eventToSend = this.client.Event(eventData);
 
-  // Send using proper transport
-  if (this.config.transport && this.config.transport === "tcp") {
-    this.client.send(eventToSend, this.client.tcp);
-  } else {
-    this.client.send(eventToSend);
+  try {
+      // Send using proper transport
```
Comment (joshgummersall) [inline lib/riemann.js]: two spaces for tabs.

File: lib/riemann.js [hunk 7]
```diff
@@ -113,11 +136,17 @@ Riemann.prototype.handlePacketEvent = function (eventString) {
...
+      if (this.config.transport && this.config.transport === "tcp") {
+        this.client.send(eventToSend, this.client.tcp);
+      } else {
+        this.client.send(eventToSend);
+      }
+  } catch (err) {
+    if (this.config.debug) {
+        console.error('Error while sending, ignoring error:', err.message);
```
Comment (joshgummersall) [inline lib/riemann.js]: two spaces for tabs.

File: lib/riemann.js [hunk 8]
```diff
@@ -9,8 +9,36 @@ var riemann = require("riemann");
...
+    } else if (this.client.tcp.socket.readyState == 'closed') {
+        if (this.config.debug) {
+            console.log('Reimann socket seems to be closed, reconnecting');
+        }
+
+        this.setupRiemannClient();
+    }
+}
+
+Riemann.prototype.setupRiemannClient = function() {
```
Comment (joshgummersall) [inline lib/riemann.js]: Including `Riemann` in the function name seems a little unnecessary, rename to `setupClient`

**Comment by joshgummersall** [comment]
@quatrix thanks a lot for submitting the patch! The functionality seems solid, the Travis build is failing because of some lint errors, you should be able to run the linter using `npm run jslint`.

**Comment by quatrix** [comment]
Applied all your comments, amended and pushed -f, please review again :)

File: lib/riemann.js [hunk 9]
```diff
@@ -18,12 +39,11 @@ function Riemann(startupTime, config, emitter) {
...
   // Bind some Riemann handlers
   this.client.once("connect", u.bind(this.onConnect, this));
-  this.client.once("error", u.bind(this.onError, this));
+  this.client.on("error", u.bind(this.onError, this));
   this.client.on("data", u.bind(this.onAck, this));
 
-  // Finally, bind to StatsD eventemitter
-  emitter.on("packet", u.bind(this.onPacket, this));
-}
+  this.emitter.on("packet", u.bind(this.onPacket, this));
```
Comment (joshgummersall) [inline lib/riemann.js:45]: I don't think you need to register events on the emitter itself on each error. This will likely double register the event handlers since the emitter is not disposed during a Riemann-level error. I would just add the emitter registration in the constructor function and leave the client event registration bits here since those will need to be re-registered each time the client is constructed.

File: lib/riemann.js [hunk 10]
```diff
@@ -113,11 +129,17 @@ Riemann.prototype.handlePacketEvent = function (eventString) {
...
-  } else {
-    this.client.send(eventToSend);
+  try {
+    // Send using proper transport
+    if (this.config.transport && this.config.transport === "tcp") {
+      this.client.send(eventToSend, this.client.tcp);
+    } else {
+      this.client.send(eventToSend);
+    }
+  } catch (err) {
```
Comment (joshgummersall) [inline lib/riemann.js:139]: So does `send` throw then if the TCP socket is closed? Or does `this.client` emit an error?

Comment (quatrix) [inline lib/riemann.js:139]: I think I added that because there were cases when send() threw an exception, and other cases when client emited an error. I wanted to just log the error and drop the event (it would be a good idea to implement some retry queue, but this is outside of the scope of this quick fix)

**Comment by joshgummersall** [comment]
@quatrix one more quick comment about calling `this.emitter.on` in the `setupClient` function. I've moved the [project over to Gitlab](https://gitlab.com/joshgummersall/statsd-riemann-backend) so I will pull your latest changes and merge through Gitlab.

File: lib/riemann.js (hunk 10 above)
Comment (joshgummersall) [inline lib/riemann.js:139]: Makes sense, just curious.

**Comment by joshgummersall** [comment]
@quatrix I've included the functionality you proposed via a Merge Request on the project's Gitlab repo. I rewrote the codebase in ES2015 last weekend so it was a bit quicker for me to simply rewrite the functionality you proposed rather than cherry picking or merging this branch in. My apologies about that, I mentioned your name and gave credit in the commit message.

Here is the merge request: https://gitlab.com/joshgummersall/statsd-riemann-backend/merge_requests/4/builds

The only two changes are [here](https://gitlab.com/joshgummersall/statsd-riemann-backend/merge_requests/4/diffs#d49744b76b1847d3816b0663abe98006ff32e5da_13_25) in that you only need to attach to `emitter` once, and also [here](https://gitlab.com/joshgummersall/statsd-riemann-backend/merge_requests/4/diffs#d49744b76b1847d3816b0663abe98006ff32e5da_118_156) where I use `onError` if an exception happens during the `send` process.

**Comment by quatrix** [comment]
thanks! :+1:
